intrinsicsolver/intrinsicsolver_test.go

```diff
@@ -0,0 +1,258 @@
+	expected := "Key: {\"Fn::Equals\" : [\"value_1\", \"value_2\"]}"
+	fixEquals(&line)
+
+	assert.Equal(t, expected, line, "MSG")
+}
+
+func TestFixUserData(t *testing.T) {
+	/*
+	   		from Doc's:
+	   		YAML:
+	   		UserData:
+	     		  Fn::Base64:
+	               !Sub |
+	                 #!/bin/bash -xe
+	                 yum update -y aws-cfn-bootstrap
+	                 /opt/aws/bin/cfn-init -v --stack ${AWS::StackName} --resource LaunchConfig --configsets wordpress_install --region ${AWS::Region}
+	                 /opt/aws/bin/cfn-signal -e $? --stack ${AWS::StackName} --resource WebServerGroup --region ${AWS::Region}
+
+	   		JSON:
+	   		UserData: { "Fn::Base64" : { "Fn::Join": ["\n", ["#!/bin/bash -xe","yum update -y aws-cfn-bootstrap",{ "Fn::Sub" : "/opt/aws/bin/cfn-init -v --stack ${AWS::StackName} --resource LaunchConfig --configsets wordpress_install --region ${AWS::Region}" },{ "Fn::Sub" : "/opt/aws/bin/cfn-signal -e $? --stack ${AWS::StackName} --resource WebServerGroup --region ${AWS::Region}" }]]}}
+
+	*/
+	line1 := "  Fn::Base64:"
+	line2 := "    !Sub |"
+	line3 := "      #!/bin/bash -xe"
+	line4 := "      yum update -y aws-cfn-bootstrap"
+	line5 := "      /opt/aws/bin/cfn-init -v --stack ${AWS::StackName} --resource LaunchConfig --configsets wordpress_install --region ${AWS::Region}"
+	line6 := "      /opt/aws/bin/cfn-signal -e $? --stack ${AWS::StackName} --resource WebServerGroup --region ${AWS::Region}"
+	line7 := ""
+	linesArray := make([]string, 0)
+	linesArray = append(linesArray, line1)
+	linesArray = append(linesArray, line2)
+	linesArray = append(linesArray, line3)
+	linesArray = append(linesArray, line4)
+	linesArray = append(linesArray, line5)
+	linesArray = append(linesArray, line6)
+	linesArray = append(linesArray, line7)
+
+	expected := "  {\"Fn::Base64\" : { \"Fn::Join\" : [\"\n\", [\"#!/bin/bash -xe\",\"yum update -y aws-cfn-bootstrap\",{ \"Fn::Sub\" : \"/opt/aws/bin/cfn-init -v --stack ${AWS::StackName} --resource LaunchConfig --configsets wordpress_install --region ${AWS::Region}\"},{ \"Fn::Sub\" : \"/opt/aws/bin/cfn-signal -e $? --stack ${AWS::StackName} --resource WebServerGroup --region ${AWS::Region}\"}]]}}"
+	fixUserData(&line1, linesArray, 0)
+
+	assert.Equal(t, expected, line1, "MSG")
+
+}
+
+func TestFixIf(t *testing.T) {
+	/*
+		   		from Doc's:
+		   		YAML:
+				   !If [condition_name, value_if_true, value_if_false]
+
+		   		JSON:
+		   		{ "Fn::If": [condition_name, value_if_true, value_if_false]}
+
+	*/
+	line1 := "!If [condition_name, value_if_true, value_if_false]"
+	line2 := ""
+	linesArray := make([]string, 0)
+	linesArray = append(linesArray, line1)
+	linesArray = append(linesArray, line2)
+
+	expected := "{ \"Fn::If\" : [\"condition_name\", \" value_if_true\", \" value_if_false\"]}"
+	fixIf(&line1, linesArray, 0)
+
+	assert.Equal(t, expected, line1, "MSG")
+}
+
+func TestFixSplit(t *testing.T) {
+	/*
+		from Doc's:
+		YAML:
+		!Split [ "|" , "a|b|c" ]
```

Comment:
Ok, I have an idea which would drastically simplify all the process and I must test if it works. Let's check what is the `goformation` behaviour in case when the YAML template has all intrinsic functions in their full form. If it would parse it with ease, there would be no need for those string equilibristics - just simplest substitutions to long forms and parsing.

Reply:
After doing some manual tests I see that only rewriting to full function forms doesn't solve the issue but doing so with expanding the YAML structure is the solution.

**This:**
````
Conditions:
  OnlyProduction: Fn::Equals: [ prod, Fn::FindInMap: [ EnvSettings, Ref: ENV, Stage ] ]
````

**Should be rewritten to this:**
````
Conditions:
  OnlyProduction: 
    Fn::Equals: 
      - prod 
      - Fn::FindInMap: 
          - EnvSettings 
          - Ref: ENV 
          - Stage
````

And the `goformation` works like a charm!
But it requires well indented YAML with homogenous notation (e.q. no curly brackets, no commas).

**This would be hipotheticaly approached by this steps:**
1. Transform []byte to string or []string array like above.
2. Do short-form-to-long-form substitution.
3. Delete all curly brackets.
4. Expand all the maps and nests into full forms (e.g. storing them somewhere, expanding, then injecting into their previous places or injecting newline marks and spaces where needed)
5. Delete all commas.

Problem could be encountered where there are very nested YAML-JSON-mixed functions properties (e.g. !If tag) which could be a pain to rewrite.
Additionally I've noticed that this whole operation preserves the UserData property from losing it's data (I have just noticed that in string-line-by-line-approach it is not well processed).

How does it sound?

Reply:
**UPDATE:**
Rewriting to that also does the trick:

````
Conditions: {'OnlyProduction': {'Fn::Equals': [prod, {'Fn::FindInMap': [EnvSettings, 'Ref': ENV, Stage]}]}}
````

Reply:
I think it will be easier to transform whole _YAML_ file to canonical form first (with curly, square brackets and quotes) then parse it and perform the replacement.

Reply:
Then I will fix it that way (this could be easier than I thought because while ago I've found out that we only need first and last curly bracket (no curly brackets inside needed but quotation marks are necessary), which additionally simplifies whole process). I'll report when I'll find the Grail.